src/mono/mono/metadata/threads.c

```diff
@@ -869,6 +868,42 @@ mono_thread_attach_internal (MonoThread *thread, gboolean force_attach, gboolean
 	return FALSE;
 }
 
+#ifndef HOST_WIN32
+static GENERATE_GET_CLASS_WITH_CACHE (wait_subsystem, "System.Threading", "WaitSubsystem");
+
+static void
+abandon_mutexes (MonoInternalThread *internal)
+{
+	ERROR_DECL (error);
+
+	MONO_STATIC_POINTER_INIT (MonoMethod, thread_exiting)
+
+		MonoClass *wait_subsystem_class = mono_class_get_wait_subsystem_class ();
+		g_assert (wait_subsystem_class);
+		thread_exiting = mono_class_get_method_from_name_checked (wait_subsystem_class, "OnThreadExiting", -1, 0, error);
+		mono_error_assert_ok (error);
+
+	MONO_STATIC_POINTER_INIT_END (MonoMethod, thread_exiting)
+
+	g_assert (thread_exiting);
+
+	if (mono_runtime_get_no_exec ())
+		return;
+
+	HANDLE_FUNCTION_ENTER ();
+
+	ERROR_DECL (error);
+
+	gpointer args [1];
+	args [0] = internal;
+	mono_runtime_try_invoke_handle (thread_exiting, NULL_HANDLE, args, error);
```

Comment:
So we saw a crash here in `Running TestUnmanagedCallersOnlyValid_OnNewNativeThread...` which is a good test - it runs some managed code on a foreign native thread.  Which means that the runtime will still have a `MonoThreadInfo` with a TLS dtor on that foreign thread which will call `mono_thread_detach_internal` (via the `thread_detach` callback) which will come here.

But by the time we get here, the domain is already unset.  So the JIT will die.

```
=================================================================
      Native stacktrace:
      =================================================================
      0x10ee5e6d6 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_dump_native_crash_info
      0x10ee024ae - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_handle_native_crash
      0x10ed5a515 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_sigsegv_signal_handler_debug
      0x7fff69677f5a - /usr/lib/system/libsystem_platform.dylib : _sigtramp
      0x0 - Unknown
      0x10ed59fb1 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : lookup_method
      0x10ed58c92 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_jit_compile_method_with_opt
      0x10ed5d09f - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_jit_runtime_invoke
      0x10ec6215f - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_runtime_try_invoke
      0x10ec643fe - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_runtime_try_invoke_handle
      0x10ec7825b - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : abandon_mutexes
      0x10ec7513c - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : mono_thread_detach_internal
      0x10ec77cdd - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : thread_detach
      0x10ece23c4 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : unregister_thread
      0x10ece2c20 - /private/tmp/helix/working/AE8409AC/p/libcoreclr.dylib : thread_info_key_dtor
      0x7fff69683163 - /usr/lib/system/libsystem_pthread.dylib : _pthread_tsd_cleanup
      0x7fff69682ee9 - /usr/lib/system/libsystem_pthread.dylib : _pthread_exit
      0x7fff6968166c - /usr/lib/system/libsystem_pthread.dylib : pthread_sigmask
      0x7fff6968150d - /usr/lib/system/libsystem_pthread.dylib : _pthread_body
      0x7fff69680bf9 - /usr/lib/system/libsystem_pthread.dylib : thread_start
 ```

---

So the gist of the current approach is: "call OnThreadExiting" from the dying thread.

I think that's not a workable idea - we already started tearing down the thread and calling managed code will be fragile at this point.

Instead I think we should do this: create a finalizable managed object that is immediately garbage and give it the responsibility to call OnThreadExiting by giving it a reference to the managed thread object.

We have the managed thread object still, because (except in the case of shutdown) it's not destroyed when the native thread is destroyed - it goes away when the last reference to it goes away - but the `MonoThreadInfo` is still holding a handle to it.

The new object will be created and it won't have any references to it as soon as the native thread exits. So on the next GC it will be garbage.  When the finalizer runs on the finalizer thread, it should abandon the mutexes.  (I guess one complication is that we can't just put a managed reference to the thread into the object since it'll be null when the finalizer runs - so maybe put a GC Handle in there - that seems dangerous, but maybe it's ok?)

Reply:
I guess we can't actually guarantee there will be a GC and that the finalizers will run